docs/source/format/Columnar.rst

```diff
@@ -1385,6 +1387,45 @@ have two entries in each RecordBatch. For a RecordBatch of this schema with
     buffer 13: col2    data
 
 
+Compression
+-----------
+
+There are three different options for compression of record batch
+body buffers: Buffers can be uncompressed, buffers can be
+compressed with the ``lz4`` compression codec, or buffers can
+be compressed with the ``zstd`` compression codec. Buffers in
+the flat sequence of a message body must be either all
+uncompressed or all compressed separately using the same codec.
+
+.. note::
+
+  ``lz4`` compression codec means the
+  `LZ4 frame format <https://github.com/lz4/lz4/blob/dev/doc/lz4_Frame_format.md>`_
+  and should not to be confused with
+  `"raw" (also called "block") format <https://github.com/lz4/lz4/blob/dev/doc/lz4_Block_format.md>`_.
+
+The difference between compressed and uncompressed buffers in the
+serialized form is as follows:
+
+* If the buffers in the :ref:`ipc-recordbatch-message` are **compressed**
+
+  - the ``data header`` includes the length and memory offset
+    of each **compressed buffer** in the record batch's body
+
+  - the ``body`` includes a flat sequence of **compressed buffers**
+    together with the **length of uncompressed buffer** as a 64-bit
+    little-endian signed integer stored in the first 8 bytes for each
+    buffer in the sequence
+
+* If the buffers in the :ref:`ipc-recordbatch-message` are **uncompressed**
+
+  - the ``data header`` includes the length and memory offset
+    of each **uncompressed buffer** in the record batch's body
+
+  - the ``body`` includes a flat sequence of **uncompressed buffers**
+    with the first 8 bytes empty or equal to ``-1`` to indicate that
+    the buffer is uncompressed
+
```

Comment:
```suggestion

.. note::

  Some Arrow implementations lack support for buffer compression. Some
  applications compress Arrow IPC streams instead of compressing buffers.
```

Reply:
@felipecrv anything else you'd suggest adding to this note?

Reply:
I would extend the note to this:

> Some Arrow implementations are not able to produce or consume IPC streams with compressed buffers in some or all buffer compression formats listed above. Applications that already make use of compression in the protocol they use to transport Arrow IPC streams SHOULD avoid double compression. Putting Arrow IPC streams with compressed buffers into another compressed data stream has very diminished returns.

Reply:
Thanks @felipecrv! I rephrased this slightly and incorporated into my suggestion above.

Reply:
@AlenkaF Let me know whether it's OK to include this note. The context behind this is that Felipe and I are working on #40465 and in HTTP services it is very common to use HTTP compression facilities to compress binary responses. We think it makes sense to add this note here to help clearly distinguish between Arrow buffer compression vs. compression applied by some other protocol to Arrow IPC data.

Reply:
What does "double compression" means here?

Reply:
@mapleFU It means "twice compressing the same data" (compressing the buffers and also compressing the stream)